lib/honeybadger/worker.rb

```diff
@@ -164,54 +171,62 @@ def run
 
     def work(msg)
       send_now(msg)
+
+      if shutdown? && throttled?
+        warn { sprintf('Unable to report %s error(s) to Honeybadger (currently throttled)', queue.size) } if queue.size > 1
+        kill!
+        return
+      end
+
       sleep(throttle_interval)
     rescue StandardError => e
       error {
         msg = "Error in worker thread class=%s message=%s\n\t%s"
         sprintf(msg, e.class, e.message.dump, Array(e.backtrace).join("\n\t"))
       }
-      sleep(1)
-    end
-
-    def throttle_interval
-      return 0 unless throttles[0]
-      mutex.synchronize do
-        throttles.reduce(1) {|a,e| a*e }
-      end
     end
 
     def notify_backend(payload)
-      debug { sprintf('worker notifying backend id=%s', payload.id) }
+      d { sprintf('worker notifying backend id=%s', payload.id) }
       backend.notify(:notices, payload)
     end
 
-    def add_throttle(t)
+    def calc_throttle_interval
+      ((BASE_THROTTLE ** throttle) - 1).round(3)
+    end
+
+    def inc_throttle
       mutex.synchronize do
-        throttles.push(t)
+        @throttle += 1
+        @throttle_interval = calc_throttle_interval
+        throttle
       end
     end
 
-    def del_throttle
+    def dec_throttle
       mutex.synchronize do
-        throttles.shift
+        return nil if throttle == 0
+        @throttle -= 1
+        @throttle_interval = calc_throttle_interval
+        throttle
       end
     end
 
     def handle_response(msg, response)
-      debug { sprintf('worker response code=%s message=%s', response.code, response.message.to_s.dump) }
+      d { sprintf('worker response code=%s message=%s', response.code, response.message.to_s.dump) }
 
       case response.code
       when 429, 503
-        warn { sprintf('Error report failed: project is sending too many errors. id=%s code=%s throttle=1.25 interval=%s', msg.id, response.code, throttle_interval) }
-        add_throttle(1.25)
+        throttle = inc_throttle
+        warn { sprintf('Error report failed: project is sending too many errors. id=%s code=%s throttle=%s interval=%s', msg.id, throttle, response.code, throttle_interval) }
```

Comment:
Looks like you have `throttle` and `response.code` flipped here.

Reply:
Heh, good catch!